### PR TITLE
fix: close watcher when FileSystemWatcher is disposed

### LINKED ISSUE #7501: Unit testing FileWatcher fails on mac, as watcher is not closed
### Bug description

When writing unit tests invloving `FileSystemWatcherImpl`, unit tests fail with the error:

```
Assertion failed: (napi_create_external(env, instance, fse_instance_destroy, NULL, &result) == napi_ok), function FSEStart, file ../src/fsevents.c, line 217.
```

It seems related to the fact that the watcher is not closed when the `FileSystemWatcherImpl` instances are disposed

### Operating system

mac os

### Installation Method

Installer from website/GitHub releases

### Version

next (development version)

### Steps to reproduce

_No response_

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by feloy**
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

Fixes #7501 

### How to test this PR?

Unit tests should pass.

Specifically, the tests in packages/main/src/plugin/filesystem-monitoring.spec.ts. 

If there is only one test in packages/main/src/plugin/filesystem-monitoring.spec.ts, the issue #7501 is not reproducible, but adding a second test (by duplicating the first one for example), makes `yarn test:unit` fail with the error:

```
Assertion failed: (napi_create_external(env, instance, fse_instance_destroy, NULL, &result) == napi_ok), function FSEStart, file ../src/fsevents.c, line 217.
```

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 1]
```diff
@@ -58,6 +60,11 @@ export class FileSystemWatcherImpl implements containerDesktopAPI.FileSystemWatc
   readonly onDidDelete: containerDesktopAPI.Event<containerDesktopAPI.Uri> = this._onDidDelete.event;
 
   dispose(): void {
+    if (this.watcher) {
+      this.watcher.close().catch((err: unknown) => {
```
Comment (feloy) [inline packages/main/src/plugin/filesystem-monitoring.ts:70]: I'm not sure if we want to do this, or if we can change the signature of the method to `async dispose(): Promise<void>`. I think that it won't respect the contract of the `Disposable` class anymore if we do this.

Comment (benoitf) [reply packages/main/src/plugin/filesystem-monitoring.ts:70]: we can't change the signature

**Review by benoitf — APPROVED** [review]

**Review by cdrage — APPROVED** [review]

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 2]
```diff
@@ -24,27 +24,40 @@ import { Disposable } from './types/disposable.js';
...
   constructor(path: string) {
     // needs to call chokidar
-    const watcher = chokidar.watch(path, {
+    this.watcher = chokidar.watch(path, {
       persistent: true,
     });
 
-    watcher.on('add', (addedPath: string) => {
+    this.watcher.on('add', (addedPath: string) => {
+      console.log('add event');
```
Comment (benoitf) [inline packages/main/src/plugin/filesystem-monitoring.ts]: Need to be removed

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 3]
```diff
@@ -24,27 +24,40 @@ import { Disposable } from './types/disposable.js';
...
+      const uri: containerDesktopAPI.Uri = Uri.file(addedPath);
+      this._onDidCreate.fire(uri);
+    });
+
+    this.watcher.on('addDir', (addedPath: string) => {
+      console.log('addDir event');
```
Comment (benoitf) [inline packages/main/src/plugin/filesystem-monitoring.ts]: to be removed

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 4]
```diff
@@ -24,27 +24,40 @@ import { Disposable } from './types/disposable.js';
...
       const uri: containerDesktopAPI.Uri = Uri.file(addedPath);
       this._onDidCreate.fire(uri);
     });
 
-    watcher.on('change', (addedPath: string) => {
+    this.watcher.on('change', (addedPath: string) => {
+      console.log('change event');
```
Comment (benoitf) [inline packages/main/src/plugin/filesystem-monitoring.ts]: remove

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 5]
```diff
@@ -24,27 +24,40 @@ import { Disposable } from './types/disposable.js';
...
       const uri: containerDesktopAPI.Uri = Uri.file(addedPath);
       this._onDidChange.fire(uri);
     });
 
-    watcher.on('unlink', (addedPath: string) => {
+    this.watcher.on('unlink', (addedPath: string) => {
+      console.log('unlink  event');
```
Comment (benoitf) [inline packages/main/src/plugin/filesystem-monitoring.ts]: Remove

File: packages/main/src/plugin/filesystem-monitoring.ts [hunk 6]
```diff
@@ -24,27 +24,40 @@ import { Disposable } from './types/disposable.js';
...
       const uri: containerDesktopAPI.Uri = Uri.file(addedPath);
       this._onDidDelete.fire(uri);
     });
 
+    console.log('max listeners', this.watcher.getMaxListeners());
```
Comment (benoitf) [inline packages/main/src/plugin/filesystem-monitoring.ts]: Remove

**Comment by feloy** [comment]
Tests constantly fail on CI. May be related to https://github.com/paulmillr/chokidar/issues/790

**Comment by feloy** [comment]
> Tests constantly fail on CI. May be related to [paulmillr/chokidar#790](https://github.com/paulmillr/chokidar/issues/790)

The latest commit added the support of the `ready` event, which fixes the problem on Windows/Linux.

But for some reason that I'm still investigating, this is not compatible with the use of the `fsevents` (by default on Mac), as no events are sent after the `ready` event (which would be expected if `persistent` was false, but is set to true).

**Comment by feloy** [comment]
> > Tests constantly fail on CI. May be related to [paulmillr/chokidar#790](https://github.com/paulmillr/chokidar/issues/790)
> 
> The latest commit added the support of the `ready` event, which fixes the problem on Windows/Linux.
> 
> But for some reason that I'm still investigating, this is not compatible with the use of the `fsevents` (by default on Mac), as no events are sent after the `ready` event (which would be expected if `persistent` was false, but is set to true).

Debugging in the chokidar sources, the problem occurs when the watched file is a symlink (see #7512). Changing the test to resolve the path, to avoid the problem

**Review by benoitf — APPROVED** [review]

**Pull request merged**